src/main/java/com/google/devtools/build/lib/bazel/repository/cache/RepoContentsCache.java

```diff
@@ -128,4 +171,86 @@ private static String getNextCounterInDir(Path entryDir) throws IOException {
       return counter;
     }
   }
+
+  public void acquireSharedLock() throws IOException {
+    Preconditions.checkState(path != null);
+    Preconditions.checkState(sharedLock == null, "this process already has the shared lock");
+    sharedLock = FileSystemLock.get(path.getRelative(LOCK_PATH), LockMode.SHARED);
+  }
+
+  public void releaseSharedLock() throws IOException {
+    Preconditions.checkState(sharedLock != null);
+    sharedLock.close();
+    sharedLock = null;
+  }
+
+  /**
+   * Creates a garbage collection {@link IdleTask} that deletes cached repos who are last accessed
+   * more than {@code maxAge} ago, with an idle delay of {@code idleDelay}.
+   */
+  public IdleTask createGcIdleTask(Duration maxAge, Duration idleDelay) {
+    Preconditions.checkState(path != null);
+    return new IdleTask() {
+      @Override
+      public String displayName() {
+        return "Repo contents cache garbage collection";
+      }
+
+      @Override
+      public Duration delay() {
+        return idleDelay;
+      }
+
+      @Override
+      public void run() throws InterruptedException, IdleTaskException {
+        try {
+          try (var lock = FileSystemLock.tryGet(path.getRelative(LOCK_PATH), LockMode.EXCLUSIVE)) {
```

Comment:
Just to confirm, the lock will only be released after `runGC` finishes? So theoretically, it could block other bazel command until it finishes? This might be annoying to users? Would it be possible to release the lock after renaming but before `deleteTree`, I guess that depends on if it's safe for multiple `deleteTree` to run at the same time?

Reply:
good idea! done. (deleting the same directory multiple times shouldn't be an issue.)